File: leetcode/leetcode_py/coin-change.py
'''
https://leetcode.com/problems/coin-change

Solution1. recursion memorization
time: o(nm) 2^n -> n
space: o(nm) height

Solution2. iteration tabulation (1D dp)
time: o(nm)
space: o(nm)
'''
class Solution:
    def coinChange_1(self, coins: List[int], amount: int) -> int:
        dp = [[-1 for i in range(amount+1)] for j in range(len(coins))]
        def findCoins(amount, idx):
            if amount <= 0 or idx == len(coins): return 0 if amount == 0 else float('inf') - 1
            if dp[idx][amount] != -1: return dp[idx][amount]

            v1 = findCoins(amount, idx+1)
            v2 = 1 + findCoins(amount-coins[idx], idx)
            dp[idx][amount] = min(v1, v2)
            return dp[idx][amount]
        res = findCoins(amount, 0)
        return res if res < float('inf') - 1 else -1

    def coinChange_2(self, coins: List[int], amount: int) -> int:
        dp = [float(inf) - 1] * (amount+1)
        dp[0] = 0
        for i in range(amount+1):
            for j, v in enumerate(coins):
                if i >= v: dp[i] = min(dp[i], 1 + dp[i - v])
        return dp[-1] if dp[-1] < float(inf) - 1 else -1
        
    def coinChange(self, coins: List[int], amount: int) -> int:
        return self.coinChange_2(coins, amount)
        

File: leetcode/leetcode_cpp/coin-change.cpp
/*
https://leetcode.com/problems/coin-change

Solution1. recursion memorization
time: o(nm) 2^n -> n
space: o(nm) height

Solution2. iteration tabulation (1D dp)
time: o(nm)
space: o(nm)
*/
class Solution {
public:
    int findCoins(vector<int>& coins, int amount, int idx, vector<vector<int>>& m) {
        if (amount <= 0 || idx == coins.size()) return amount == 0 ? 0 : INT_MAX-1;
        if (m[idx][amount] != -1) return m[idx][amount];

        int v1 = findCoins(coins, amount, idx+1, m); // move to next coin
        int v2 = 1 + findCoins(coins, amount - coins[idx], idx, m); // take this coin
        m[idx][amount] = min(v1, v2);
        return m[idx][amount];
    }
    int coinChange_1(vector<int>& coins, int amount) {
        vector<vector<int>> m(coins.size(), vector<int>(amount+1, -1)); // [coinIdx][amount]
        int res = findCoins(coins, amount, 0, m);
        return res < INT_MAX - 1 ? res : -1;                
    }
    int coinChange_2(vector<int>& coins, int amount) {
        vector<int> dp(amount+1, INT_MAX-1);
        dp[0] = 0;
        for (int i=1; i<= amount; i++) {
            for (int j=0; j<coins.size(); ++j) {
                if (i >= coins[j]) dp[i] = min(dp[i], 1 + dp[i - coins[j]]);
            }
        }
        return dp[amount] < INT_MAX - 1 ? dp[amount] : -1;
    }
    int coinChange(vector<int>& coins, int amount) {
        return coinChange_2(coins, amount);
    }
};